### ISSUE TITLE
[Do It! 딥러닝 프로그래밍 입문] Gradient Clipping에 대한 시각적인 이해를 돕는 자료 정리

### ISSUE THREAD

**Issue by RyanKor** [RyanKor/dl-ml-book-code#2]
처음에 딥러닝 프로그래밍 책을 볼 때 코세라에서 다중 분류 모델 코드를 짤 때 못봤던 `np.clip`이라는 메소드를 사용했고, 이게 도대체 뭔데 이렇게 자주 쓰나 싶었는데, 오늘 코세라 RNN 수업 내용을 돌아보다가 관련 내용을 찾게 되어서 한 번 더 정리하게 되었다.

딥러닝 프로그래밍 책에선 아래와 같이 fit 이나 forward & backward propagation을 사용할 때도 `np.clip`을 사용하는데 하단의 이미지를 보면 이걸 왜 사용하는지 직관적으로 이해가 된다.

```python
    def fit(self, x, y, epochs=100, x_val=None, y_val=None):
        
        np.random.seed(42)
        
        self.init_weights(x.shape[1], y.shape[1])
        
        for i in range(epochs):
            
            loss = 0
            
            print('.', end="")
            
            for x_batch, y_batch in self.gen_batch(x,y):
                
                a = self.training(x_batch, y_batch)
                
                a = np.clip(a, 1e-10, 1-1e-10)
                
                loss += np.sum(-y_batch * np.log(a))
                
            
            self.losses.append((loss + self.reg_loss()) / len(x))
            
            self.update_val_loss(x_val, y_val)
```

![image](https://user-images.githubusercontent.com/40455392/137669606-b222ffb1-b899-4ae3-9bbd-93c5cc527d4a.png)

막연하게 가중치 값을 너무 튀지 않게끔만 잡아준다고 생각했는데, 그것 이상의 큰 역할이 있었다.

최적의 학습률을 보이는 방향으로 가게끔 학습율을 조절해주듯 적절하게 기울기를 조절해주는 것이다.

아마 딥러닝 프로그래밍 책에서 learning_rate를 적용하는 시점에서도 항상 `np.clip`을 사용했던 것을 생각해보면 두 개를 함께 사용해도 큰 문제는 없어 보이니, 적절하게 사용하면 좋을 것 같다.
